packages/block/block.ts

```diff
@@ -0,0 +1,212 @@
+          setAttribute(el, edit.name, value);
+        }
+      }
+    }
+
+    this.el = root;
+    if (parent) insertBefore$.call(parent, root, refNode);
+
+    return root;
+  }
+  patch(block: AbstractBlock): HTMLElement {
+    const root = this.el as HTMLElement;
+    if (!block.props) return root;
+    const props = this.props!;
+    // If props are the same, no need to patch
+    if (!this.shouldUpdate(props, block.props)) return root;
+    this.props = block.props;
+
+    for (let i = 0, j = this.edits.length; i < j; ++i) {
+      const current = this.edits[i]!;
+      const el = getCurrentElement(current, root, this.cache, i);
+      for (let k = 0, l = current.edits.length; k < l; ++k) {
+        const edit = current.edits[k]!;
+        if (edit.type === EditType.Block) {
+          edit.block.patch(block.edits?.[i]![k].block);
+          continue;
+        }
+        if (!('hole' in edit) || !edit.hole) continue;
+        const oldValue = props[edit.hole];
+        const newValue = block.props[edit.hole];
+
+        if (newValue === oldValue) continue;
+
+        if (edit.type === EditType.Event) {
+          edit.patch?.(newValue);
+        } else if (edit.type === EditType.Child) {
+          if (oldValue instanceof AbstractBlock) {
+            // Remember! If we find a block inside a child, we need to locate
+            // the cooresponding block in the new props and patch it.
+            const firstEdit = block.edits?.[i]?.edits[k] as EditChild;
+            const thisSubBlock = block.props[firstEdit.hole];
+            oldValue.patch(thisSubBlock);
+            continue;
+          }
+          setText(el, String(newValue), edit.index);
+        } else {
+          setAttribute(el, edit.name, newValue);
+        }
+      }
+    }
+
+    return root;
+  }
+  move(block: AbstractBlock | null = null, refNode: Node | null = null) {
+    insertBefore$.call(this.parent, this.el!, block ? block.el! : refNode);
+  }
+  remove() {
+    removeElement$.call(this.el);
+  }
+  shouldUpdate(oldProps: Props, newProps: Props): boolean {
+    for (const i in oldProps) {
+      if (oldProps[i] !== newProps[i]) return true;
+    }
+    return false;
+  }
+  toString() {
+    return String(this.el?.outerHTML);
+  }
+  get parent(): HTMLElement | null | undefined {
+    if (!this._parent) this._parent = this.el?.parentElement;
+    return this._parent;
+  }
+}
+
+const getCurrentElement = (
+  current: Edit,
+  root: HTMLElement,
+  cache?: Map<number, HTMLElement>,
+  slot?: number, // edit index
+): HTMLElement => {
+  if (!current.path.length) return root;
+  if (cache && slot !== undefined && mapHas$.call(cache, slot)) {
+    return mapGet$.call(cache, slot)!;
+  }
+  // path is an array of indices to traverse the DOM tree
+  // For example, [0, 1, 2] becomes root.childNodes[0].childNodes[1].childNodes[2]
+  // We use path because we don't have the actual DOM nodes until mount()
+  for (let k = 0, l = current.path.length; k < l; ++k) {
```

Comment:
```suggestion
  const l = current.path.length;
  if (!l) return root;
  if (cache && slot !== undefined && mapHas$.call(cache, slot)) {
    return mapGet$.call(cache, slot)!;
  }
  // path is an array of indices to traverse the DOM tree
  // For example, [0, 1, 2] becomes root.childNodes[0].childNodes[1].childNodes[2]
  // We use path because we don't have the actual DOM nodes until mount()
  for (let k = 0; k < l; ++k) {
```